sklearn/model_selection/_search.py

```diff
@@ -557,54 +560,146 @@ def _fit(self, X, y, labels, parameter_iterable):
-                score += this_score
             if self.iid:
-                score /= float(n_test_samples)
+                mean_score = (test_scores * test_sample_counts).sum()
+                mean_score /= test_sample_counts.sum()
             else:
-                score /= float(n_splits)
-            scores.append((score, parameters))
-            # TODO: shall we also store the test_fold_sizes?
-            grid_scores.append(_CVScoreTuple(
-                parameters,
-                score,
-                np.array(all_scores)))
-        # Store the computed scores
-        self.grid_scores_ = grid_scores
+                mean_score = test_scores.mean()
+
+            for split_i in range(n_splits):
+                results["%s_test_split_%s" % (self.scorer_name_, split_i)][
+                    candidate_i] = test_scores[split_i]
+
+            # Store the mean score and the parameters for this fit
+            results["%s_test_mean" % scorer_name][candidate_i] = mean_score
+            results["%s_test_std" % self.scorer_name_][
+                candidate_i] = test_scores.std()
+
+        candidate_dicts = all_parameters[::n_splits]
+        for candidate_i in range(self.n_candidates_):
+            for param, value in six.iteritems(candidate_dicts[candidate_i]):
+                # This entry alone gets unmasked when assigned
+                results["param_%s" % param][candidate_i] = value
 
         # Find the best parameters by comparing on the mean validation score:
-        # note that `sorted` is deterministic in the way it breaks ties
-        best = sorted(grid_scores, key=lambda x: x.mean_validation_score,
-                      reverse=True)[0]
-        self.best_params_ = best.parameters
-        self.best_score_ = best.mean_validation_score
+        # note that `sorted` is preferred over ``np.argsort`` as it
+        # deterministically selects the candidate with the least param value
+        sorted_indices = sorted(
+            range(self.n_candidates_),
+            key=results["%s_test_mean" % scorer_name].__getitem__,
+            reverse=True)
+
+        results["%s_test_rank" % scorer_name][sorted_indices] = (
+            np.arange(1, self.n_candidates_ + 1))
+
+        self.results_ = results
+        self.best_index_ = sorted_indices[0]
+        best_parameters = self._get_candidate_params(self.best_index_)
 
         if self.refit:
             # fit the best estimator using the entire dataset
             # clone first to work around broken estimators
             best_estimator = clone(base_estimator).set_params(
-                **best.parameters)
+                **best_parameters)
             if y is not None:
                 best_estimator.fit(X, y, **self.fit_params)
             else:
                 best_estimator.fit(X, **self.fit_params)
             self.best_estimator_ = best_estimator
         return self
 
+    def _get_candidate_params(self, index):
```

Comment:
I still think we should redundantly store a parameter dict for each candidate. Much easier to use straight out, even if it's also valuable to have each parameter as a separate column.